marshaling_rules.md

```diff
@@ -0,0 +1,165 @@
+
+- `Numerics` is a discriminated union of `Integers | Doubles`. Accepts all numeric inputs, but leaves it up to the user to decipher what exactly was received from `R`. No runtime validation, no extra allocation, ALTREPS remain unfolded. 
+- `ComplexNumerics` represents either `Complexes` or `Numerics`
+
+
+----------------------------------------------------------------------------
+
+# Underlying vector types
+## Terminology
+A 'vector' is a primitive type used in `R`. Vectors are designed to behave as a strongly typed 1D array of objects. There are two different implementations of vector types: one is basically a pointer to a contiguous block of memory with known length (and some additional metadata), another is an iterator deigned to store rules for generating sequences of elements (instead of storing potentially very large vectors in memory). Array-based vectors shall be referred to as 'plain old data' (POD), iterators -- as ALTREP.
+
+`R` recognizes the following vector types that are directly exposed to the user:
+ - `logical (i32)`
+ - `integer (i32)`
+ - `real (f64)`
+ - `complex (f64, f64)`
+ - `raw (u8)`
+ - `character (usize)` (collection of pointers to character arrays)
+
+Each vector can contain special `NA` values. None of the primitive types have built-in support for `NA` (including `f64`, which has notion of `NaN`, a different thing), so `R` treats one value from the range of allowed values as `NA`. For instance, `NA_integer_` is `i32::MIN`, which is `1i32 << 31 = -2147483648`. As a result, `x <- -2147483648L` results in an error in `R`.
+
+## `Rust` counterparts
+
+`R` objects passed to `Rust` require additional validation and transformation. Let us define the following types:
+- `struct Rint(i32)`
+- `struct Rbool(i32)`
+- `struct Rfloat(f64)`
+- `struct Rbyte(u8)`
+- `struct Rcomplex((f64, f64))`
+- `struct Rstr(usize)` (?)
+
+Note: `complex` is an `(f64, f64)` struct
+
+Each of these types is binary compatible with their underlying type. An array of, say `i32`, represented by a `*i32` pointer and length, can be viewed as `*Rint` of the same length. 
+This can be the preferred solution when dealing with `R` plain vectors.
+
+For each supported primitive type `T` `Rt` would be its minimal wrapper. E.g., for `T = i32`, `Rt = Rint`.
+`extendr` prefers `Rt` over `T` types. Parameters that use `T`-derived types will require runtime `NA` validation, introducing implicit overhead.
+
+Type conversion traits for `Rt` are:
+- `Into<Option<T>>` (this is always a valid conversion)
+- `TryInto<T>`, errors on `NA`
+- `TryFrom<T>`, errors if provided argument equals to the value reserved for `NA`
+- `TryFrom<Option<T>>` for the same reason, as `Some(i32::MIN)` is invalid `Rint` value (instead, `None` should be used)
```

Comment:
I don't entirely understand what you're trying to say here about `TryFrom<Option<T>>`. For the same reason what? Could you add 1-2 sentences to explain?

Reply:
Done, rephrased it in the last commit.

Reply:
Thanks, I get it now.